### PR TITLE
Augment Functionality Infer() of KNN Classifier

### LINKED ISSUE #2285: Augment Infer() of KNN Classifier

### PR DESCRIPTION
**Pull request by mihail911**
These fixes address #2271, augmenting the infer() method of KNN Classifier to return top 'n' most frequent classes. Fixes #2285. Let me know if I need to change anything. @BoltzmannBrain @rmccall84 @scottpurdy

### CODE REVIEW

**Comment by BoltzmannBrain** [comment]
@mihail911 please add error messages to your assert statements; otherwise it's difficult to tell what's failing.

File: nupic/algorithms/KNNClassifier.py [hunk 1]
```diff
@@ -573,38 +578,32 @@ def getDistances(self, inputPattern):
...
-                        are in that category.
-      dist:             A list of length numPrototypes. Each entry is the
-                        distance from the unknown to that prototype. All
-                        distances are between 0.0 and 1.0
-      categoryDist:     A list of length numCategories. Each entry is the
-                        distance from the unknown to the nearest prototype of
-                        that category. All distances are between 0 and 1.0.
+  ##########################################################################
+  def infer(self, inputPattern, numWinners=1, computeScores=True,
+                  overCategories=True, partitionId=None):
```
Comment (breznak) [inline nupic/algorithms/KNNClassifier.py]: nit: maybe use one per line param to look better?

File: nupic/algorithms/KNNClassifier.py [hunk 2]
```diff
@@ -573,38 +578,32 @@ def getDistances(self, inputPattern):
...
+    """Find the category that best matches the input pattern. Returns the
+    winning category `numWinners` categories indices as an array plus  a
+    distribution over all categories.
+
+    This method returns a 4 item tuple:
+    (winner, inferenceResult, dist, categoryDist)
+      winners: The `numWinners` categories with the greatest number of nearest neighbors within
```
Comment (breznak) [inline nupic/algorithms/KNNClassifier.py]: please convert to `@param winners: ....` docstring format

File: nupic/algorithms/KNNClassifier.py [hunk 3]
```diff
@@ -573,38 +578,32 @@ def getDistances(self, inputPattern):
...
-                        that category. All distances are between 0 and 1.0.
+  ##########################################################################
+  def infer(self, inputPattern, numWinners=1, computeScores=True,
+                  overCategories=True, partitionId=None):
+    """Find the category that best matches the input pattern. Returns the
+    winning category `numWinners` categories indices as an array plus  a
+    distribution over all categories.
+
+    This method returns a 4 item tuple:
+    (winner, inferenceResult, dist, categoryDist)
```
Comment (breznak) [inline nupic/algorithms/KNNClassifier.py]: ditto, `@return ....`

File: nupic/algorithms/KNNClassifier.py (hunk 1 above)
Comment (scottpurdy) [inline nupic/algorithms/KNNClassifier.py]: Either is fine. Examples in PEP-8 use this style. But make sure to line up indentation correctly (see [PEP-8](https://www.python.org/dev/peps/pep-0008/#indentation))

File: nupic/algorithms/KNNClassifier.py [hunk 4]
```diff
@@ -539,6 +540,10 @@ def learn(self, inputPattern, inputCategory, partitionId=None, isSparse=0,
           and self._numPatterns == self.numSVDSamples:
         self.computeSVD()
 
+    #print "Category List: ", self._categoryList
```
Comment (scottpurdy) [inline nupic/algorithms/KNNClassifier.py]: remove these?

File: nupic/algorithms/KNNClassifier.py [hunk 5]
```diff
@@ -629,10 +628,11 @@ def infer(self, inputPattern, computeScores=True, overCategories=True,
 
       # Prepare inference results.
       if inferenceResult.any():
-        winner = inferenceResult.argmax()
+        #Perform argsort in descending order and return top `n`
```
Comment (scottpurdy) [inline nupic/algorithms/KNNClassifier.py]: Space after the `#`

File: nupic/algorithms/KNNClassifier.py (hunk 2 above)
Comment (scottpurdy) [inline nupic/algorithms/KNNClassifier.py]: I am confused by this. What happens with these categories? Are they in the return tuple? That says "winner" so I don't see what you would do with multiple.

File: nupic/algorithms/KNNClassifier.py [hunk 6]
```diff
@@ -573,38 +578,32 @@ def getDistances(self, inputPattern):
...
-                        number of neighbors within the top k neighbors that
-                        are in that category.
-      dist:             A list of length numPrototypes. Each entry is the
-                        distance from the unknown to that prototype. All
-                        distances are between 0.0 and 1.0
-      categoryDist:     A list of length numCategories. Each entry is the
-                        distance from the unknown to the nearest prototype of
-                        that category. All distances are between 0 and 1.0.
+  ##########################################################################
+  def infer(self, inputPattern, numWinners=1, computeScores=True,
```
Comment (scottpurdy) [inline nupic/algorithms/KNNClassifier.py]: Might be better to put `numWinners` at the end so we don't break any current clients that use positional arguments.

Comment (scottpurdy) [inline nupic/algorithms/KNNClassifier.py]: That said, if it makes much more sense to have earlier in the list that is ok

File: nupic/algorithms/KNNClassifier.py [hunk 7]
```diff
@@ -641,12 +641,23 @@ def infer(self, inputPattern, computeScores=True, overCategories=True,
...
       print "  pct neighbors of each category:", inferenceResult
       print "  dist of each prototype:", dist
       print "  dist of each category:", categoryDist
 
-    result = (winner, inferenceResult, dist, categoryDist)
+    result = (winners, inferenceResult, dist, categoryDist)
+    numCategories = len(self._categoryList)
+
+    if numWinners > numCategories:
+          warnings.warn("Your requested number of categories exceeds the "
```
Comment (BoltzmannBrain) [inline nupic/algorithms/KNNClassifier.py]: indents = 2 spaces

File: nupic/algorithms/KNNClassifier.py [hunk 8]
```diff
@@ -641,12 +641,23 @@ def infer(self, inputPattern, computeScores=True, overCategories=True,
...
+                         "total number of categories learned. Returning %d "
```
Comment (BoltzmannBrain) [inline nupic/algorithms/KNNClassifier.py]: `numCategories` should always be type int, so use `%i`

File: nupic/algorithms/KNNClassifier.py [hunk 9]
```diff
@@ -641,12 +641,23 @@ def infer(self, inputPattern, computeScores=True, overCategories=True,
...
+                         "categories..." %(numCategories))
+    if numWinners > self.k:
```
Comment (BoltzmannBrain) [inline nupic/algorithms/KNNClassifier.py]: This shouldn't make a difference. If `k` = 1, we still want a list of `numWinners` categories. The param `k` is used to calculate distances to potential classification categories. @mihail911 I can explain this further in person if you'd like.

**Comment by mihail911** [comment]
Will return to this after initial fix to classification base class in nupic.fluent is made
@scottpurdy @BoltzmannBrain @subutai

**Pull request closed without merging**